polar-core/src/filter.rs

```diff
@@ -0,0 +1,491 @@
+use std::{
+    collections::{HashMap, HashSet},
+    fmt::{self, Display, Formatter},
+    hash::Hash,
+};
+
+use crate::{
+    data_filtering::{unregistered_field_error, unsupported_op_error, PartialResults, Type, Types},
+    error::{invalid_state_error, PolarResult, RuntimeError},
+    events::ResultEvent,
+    terms::*,
+};
+
+use serde::Serialize;
+
+type TypeName = String;
+type FieldName = String;
+type Map<A, B> = HashMap<A, B>;
+type Set<A> = HashSet<A>;
+
+#[derive(Clone, Eq, Debug, Serialize, PartialEq)]
+pub struct Filter {
+    root: TypeName,                  // the host already has this, so we could leave it off
+    conditions: Vec<Set<Condition>>, // disjunctive normal form
+    relations: Set<Relation>,
+}
+
+#[derive(PartialEq, Eq, Debug, Serialize, Clone, Hash)]
+pub struct Relation(TypeName, FieldName, TypeName);
```

Comment:
I just hate how using struct variants blows up the code so much.

Reply:
In the definition? Or on field access? Or matches?

Reply:
All those three and also when you construct it

Reply:
See my comment below: https://github.com/osohq/oso/pull/1308#discussion_r751727802

I feel that having field names so that the code that accesses fields is easy to understand (`rel.through` instead of `rel.1`)  outweighs the space saving of using tuple style structs in matches or construction.

With `rel.1` I have to remember what that means.

Reply:
Here's another way to fix that though if you don't want to add names: https://github.com/osohq/oso/pull/1308/files/e142e9b86b85b531d0bb74eb558dffb51d17149c#r753503409

I think if you used that style everywhere when accessing fields it would have the same benefit.

Reply:
(But note that that's harder to enforce going forward.) With names you or anyone else in the future has to use the name to pull the field out of the struct. Helps us maintain things a bit more easily.

Reply:
Oy vey Haskell handles this so much better

Reply:
O how